src/recompiler.ts

```diff
@@ -0,0 +1,203 @@
+import fs from "fs";
+import path from "path";
+import { createHash } from "crypto";
+import { HardhatRuntimeEnvironment, TaskArguments } from "hardhat/types";
+import { starknetCompileAction } from "./task-actions";
+import { getArtifactPath, traverseFiles } from "./utils";
+import { ABI_SUFFIX } from "./constants";
+
+const fsPromises = fs.promises;
+
+interface ContractData {
+    contentHash: string;
+    outputPath: string;
+    abiPath: string;
+}
+
+// FileName - HashPair
+interface Cache {
+    [key: string]: ContractData;
+}
+
+// Cache file name
+const CACHE_FILE_NAME = "cairo-files-cache.json";
+
+// Creates cache file only if it doesn't exist in cache
+const upsertFile = async (cacheDirName: string): Promise<Cache> => {
+    const cacheFile = path.join(cacheDirName, CACHE_FILE_NAME);
+    const cacheDirpath = path.join(cacheDirName);
+
+    // Creates cache directory if it doesn't exist
+    if (!fs.existsSync(cacheDirpath)) {
+        fs.mkdirSync(cacheDirpath, { recursive: true });
+    }
+
+    if (!fs.existsSync(cacheFile)) {
+        // create file, if it's not found
+        await fsPromises.writeFile(cacheFile, JSON.stringify({}));
+        return {};
+    } else {
+        // try to read file
+        const oldFile = await fsPromises.readFile(cacheFile);
+        const oldCacheEntry: Cache = JSON.parse(oldFile.toString() || "{}");
+        return oldCacheEntry;
+    }
+};
+
+// Gets hash of each .cairo file inside contracts
+const getContractHash = async (hre: HardhatRuntimeEnvironment): Promise<Cache> => {
+    const { starknetSources: defaultSourcesPath } = hre.config.paths;
```

Comment:
This is the biggest issue with recompilation I believe. Because the user can also specify `--account-contract` and `disable-hint-validation` flags (not to speak of using different compiler versions).

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/106/commits/fe5195a7d3efb962794159135f3dd99b0a17da0d